### PR TITLE
PoissonVI to model quantitative scATAC-seq fragment counts

### LINKED ISSUE #1926: Implement PoissonVI to model scATAC-seq fragment counts
**Is your feature request related to a problem? Please describe.**
PeakVI currently only supports modelling of binarized scATAC-seq data. In our [preprint](https://www.biorxiv.org/content/10.1101/2022.05.04.490536v1), we showed that it can be beneficial to actually model the scATAC-seq fragment counts.

**Describe the solution you'd like**
As previously discussed with @adamgayoso, it would be good to add the PoissonVI model to scvi, that users can also model the fragment count data.

### PR DESCRIPTION
**Pull request by lauradmartens**
-   [x] Closes #1926 
-   [x] Tests added and passed if fixing a bug or adding a new feature
      - Where should I add the test for the Layer field?
-   [x] All code checks passed.
      -  Fails with Python 3.9 because of typing error
-   [x] Added type annotations to new arguments/methods/functions.
-   [x] Added an entry in the latest `docs/release_notes/index.md` file if fixing a bug or adding a new feature.

## New things added with PoissonVI:
- New LayerField `FragmentCountLayerField` that checks that the provided matrix does not contain binarized scATAC counts or read counts (which normally have more even than uneven counts)
- PoissonVI model that models fragment counts with a Poisson distribution. This used the default `VAE` model but differs from scvi because it: 
     - uses the default PeakVI architecture and training procedure
     - Implements the functions `get_accessibility_estimates` and `differential_accessibility`
     - In `get_accessibility_estimates` one has the option to compute normalized accessibility estimates without the region specific bias which is essentially the bias parameter of the last layer of the model (similar to PeakVi's region parameter)

### CODE REVIEW

File: scvi/external/poissonvi/_model.py
Comment (martinkim0) [inline scvi/external/poissonvi/_model.py:1]:
```suggestion
from __future__ import annotations

import logging
```
I think this will fix the typing error in Python 3.9. I'll take a look at the rest of the code later. Thanks!

Comment (lauradmartens) [reply scvi/external/poissonvi/_model.py:3]: Thanks!

File: scvi/data/_utils.py
Comment (martinkim0) [inline scvi/data/_utils.py:353]: I think it would be preferable to use pure numpy here instead of jax because, for most use cases, this function will only be called once. Jax jit is generally not worth it unless the function is called multiple times.

File: scvi/data/_utils.py
Comment (martinkim0) [inline scvi/data/_utils.py:333]: Cleaner to use `scipy.sparse.issparse`

File: scvi/data/_utils.py
Comment (martinkim0) [inline scvi/data/_utils.py:324]: Need to check for the unlikely edge case where `adata.n_obs < n_to_check`.

File: scvi/data/_utils.py
Comment (martinkim0) [inline scvi/data/_utils.py:329]: Same here

File: scvi/data/_utils.py
Comment (martinkim0) [inline scvi/data/_utils.py:336]: This line could potentially be copying the entire dataframe into a numpy. In terms of performance, it would be preferable to subset the dataframe first and then convert to numpy.

File: scvi/data/_utils.py
Comment (martinkim0) [reply scvi/data/_utils.py:353]: I'd also just refactor so that the logic in `_is_not_fragment_val` is inside `_check_fragment_counts`

File: scvi/data/fields/_layer_field.py
Comment (martinkim0) [inline scvi/data/fields/_layer_field.py:184]: Since this new class only overwrites `validate_field`, it would be cleaner to instead include an argument in `LayerField.__init__` (_e.g._ `is_fragment_count_data`) and then add the appropriate check into `LayerField.validate_field`

File: tests/external/test_poissonvi.py
Comment (martinkim0) [inline tests/external/test_poissonvi.py:14]:
```suggestion
```

File: tests/external/test_poissonvi.py
Comment (martinkim0) [inline tests/external/test_poissonvi.py:13]:
```suggestion
    adata = synthetic_iid(batch_size=100)
```

File: tests/external/test_poissonvi.py
Comment (martinkim0) [inline tests/external/test_poissonvi.py:11]:
```suggestion
```

**Review by martinkim0 — COMMENTED** [review]
Thanks for the contribution! One comment that I wasn't able to include: since `POISSONVI` needs to reproduce the default parameters of `PEAKVI`, it would be preferable if the latter is a subclass of the former so there is less repeated code. And then you could add a parameter `likelihood` to the model init that is defaulted to `"poisson"`.

**Comment by lauradmartens** [comment]
Thanks a lot for the comments! I addressed the minor changes on the `_check_fragment_counts` and will rewrite the model as a subclass of `PEAKVI` by the end of the week.

**Comment by lauradmartens** [comment]
Thanks again for your comments! I thought again of the suggestions of making `PEAKVI` a subclass of `POISSONVI` to have less repeated code for the model architecture. If I understood it correctly, you would call `PEAKVI` init with an additional likelihood parameter set to `"poisson"`? I am unsure how to make this work with `VAE` as the module class, because `PEAKVI/PEAKVAE` has many arguments that are not in `VAE` (`n_input_regions`, `region_factors`, `model_depth`).

**Comment by martinkim0** [comment]
Hi @lauradmartens, sorry for the late reply. My understanding is that `POISSONVI` uses the default model architecture of `PEAKVI`, so maybe it makes more sense to use `PEAKVAE` as the module class instead of `VAE`? This would make it easier to subclass from `PEAKVI` instead. Let me know what you think!

**Comment by lauradmartens** [comment]
Hi @martinkim0! I think it's actually easier to use the `VAE` model because it implements all the functions to deal with Poisson/negative binomial loss whereas the `PEAKVAE` model has the binary decoder and the additional region and cell-specific factors. When I said the same architecture, I just meant the same number of layers, etc. to deal with the much larger feature space in scATAC-seq compared to scRNA-seq. I now wrote it in a way that `POISSONVI` inherits from `PEAKVI`, which reduces the doubled code at least for the training. Let me know what you think :)

File: scvi/external/poissonvi/_model.py
Comment (martinkim0) [inline scvi/external/poissonvi/_model.py:77]: Needs to be addressed

File: scvi/external/poissonvi/_model.py
Comment (martinkim0) [inline scvi/external/poissonvi/_model.py:246]: Preferable to raise a `NotImplementedError` here

File: scvi/data/fields/_layer_field.py
Comment (martinkim0) [inline scvi/data/fields/_layer_field.py:184]: Taking a second look at this, I think it would be cleaner if we're able to directly modify `LayerField` instead. In other words, we can add an argument `check_fragment_counts` to `LayerField.__init__` and then have 
```
if self.check_fragment_counts and not _check_fragment_counts(x):
    ....
```
in `LayerField.validate_field`. What do you think?

File: scvi/external/poissonvi/_model.py
Comment (martinkim0) [inline scvi/external/poissonvi/_model.py:393]: `NotImplementedError` here too

**Review by martinkim0 — COMMENTED** [review]
Looks great overall! Left some minor comments and it should be good to merge after that. Let me know if you're also interested in contributing a tutorial

**Comment by lauradmartens** [comment]
Hi Martin! I added the changes as you suggested and used a provisional link for the tutorial. I am of course happy to also contribute a tutorial. Would you also like me to do a `user_guide` section of what is the difference?

**Comment by martinkim0** [comment]
Thanks for addressing the comments! The user guide is optional - it should just be a condensed explanation of how the model works. Since we already have a PeakVI [one](https://docs.scvi-tools.org/en/stable/user_guide/models/peakvi.html), you could just copy paste and make the necessary changes. For tutorials, you can put in a PR in our [tutorials repo](https://github.com/scverse/scvi-tutorials).

**Review by martinkim0 — APPROVED** [review]
LGTM

**Comment by martinkim0** [comment]
Hi @lauradmartens, what's the status on this? Are there any changes you wanted to add before merging?

**Comment by lauradmartens** [comment]
No it looks good to me. I will add the tutorial by the end of the week.

**Comment by martinkim0** [comment]
Hey @lauradmartens! Thanks for your work so far - I'm planning on releasing this with scvi-tools 1.1.0 within a couple of weeks. I will probably post a Twitter thread with release updates. I was wondering if you'd like me to tag you in the thread (if so, please reply with your handle) and/or if you'd like to write the tweet describing the model (if so, please reply with the summary). Let me know!

**Comment by lauradmartens** [comment]
Thanks @martinkim0! My twitter handle is `lauradmartens`.

For the description maybe something like:
Modeling quantitative scATAC-seq fragment counts is now possible using the [PoissonVI model](https://www.nature.com/articles/s41592-023-02112-6). Also check out the tutorial on how you can convert your 10x matrices to fragment count matrices.

**Pull request merged**
